File: backend/internal/router/plugin_engine.go
// Copyright (c) 2026 FayHub Team
// SPDX-License-Identifier: MIT

package router

import (
	"fayhub/internal/controller"
	"fayhub/internal/middleware"

	"github.com/gin-gonic/gin"
)

type PluginEngineRouter struct{}

func (r *PluginEngineRouter) Init(router *gin.Engine) {
	pluginEngineGroup := router.Group("/api/plugin-engine")
	pluginEngineGroup.Use(middleware.JwtAuthMiddleware())
	pluginEngineGroup.Use(middleware.TenantMiddleware())
	{
		// 插件管理接口（需要租户权限）
		pluginEngineGroup.GET("/plugins", controller.ControllerGroupApp.PluginEngineController.ListPlugins)
		pluginEngineGroup.GET("/plugins/:id", controller.ControllerGroupApp.PluginEngineController.GetPlugin)
		pluginEngineGroup.DELETE("/plugins/:id", controller.ControllerGroupApp.PluginEngineController.UninstallPlugin)
		pluginEngineGroup.PUT("/plugins/:id/enable", controller.ControllerGroupApp.PluginEngineController.EnablePlugin)
		pluginEngineGroup.PUT("/plugins/:id/disable", controller.ControllerGroupApp.PluginEngineController.DisablePlugin)
		pluginEngineGroup.PUT("/plugins/:id/upgrade", controller.ControllerGroupApp.PluginEngineController.UpgradePlugin)
		pluginEngineGroup.GET("/plugins/:id/config", controller.ControllerGroupApp.PluginEngineController.GetPluginConfig)
		pluginEngineGroup.PUT("/plugins/:id/config", controller.ControllerGroupApp.PluginEngineController.UpdatePluginConfig)
		pluginEngineGroup.GET("/plugins/:id/page", controller.ControllerGroupApp.PluginEngineController.GetPluginPage)

		pluginEngineGroup.GET("/plugins/:id/versions", controller.ControllerGroupApp.PluginEngineController.GetVersionHistory)
		pluginEngineGroup.POST("/plugins/:id/rollback", controller.ControllerGroupApp.PluginEngineController.RollbackPlugin)
		pluginEngineGroup.GET("/plugins/:id/check-update", controller.ControllerGroupApp.PluginEngineController.CheckForUpdates)
		pluginEngineGroup.GET("/check-updates", controller.ControllerGroupApp.PluginEngineController.CheckAllUpdates)

		pluginEngineGroup.GET("/plugins/:id/dependencies", controller.ControllerGroupApp.PluginEngineController.GetDependencies)
		pluginEngineGroup.POST("/plugins/:id/dependencies", controller.ControllerGroupApp.PluginEngineController.SaveDependencies)
		pluginEngineGroup.GET("/plugins/:id/validate-deps", controller.ControllerGroupApp.PluginEngineController.ValidateDependencies)
	}

	pluginEngineGroup.POST("/install-callback", controller.ControllerGroupApp.PluginEngineController.InstallCallback)
	pluginEngineGroup.POST("/demo/install", controller.ControllerGroupApp.PluginEngineController.InstallDemoPlugin)

	marketGroup := router.Group("/api/plugin-engine/market")
	marketGroup.Use(middleware.JwtAuthMiddleware())
	marketGroup.Use(middleware.TenantMiddleware())
	{
		marketGroup.GET("/search", controller.ControllerGroupApp.PluginEngineController.SearchMarketPlugins)
		marketGroup.GET("/plugins/:id", controller.ControllerGroupApp.PluginEngineController.GetMarketPluginDetail)
		marketGroup.GET("/categories", controller.ControllerGroupApp.PluginEngineController.GetMarketCategories)
		marketGroup.POST("/install", controller.ControllerGroupApp.PluginEngineController.InstallFromMarket)
		marketGroup.GET("/public-key", controller.ControllerGroupApp.PluginEngineController.GetMarketPublicKey)
	}

	router.GET("/plugin-assets/:pluginId/*filepath", controller.ControllerGroupApp.PluginEngineController.ServePluginAsset)

	pluginDataGroup := router.Group("/api/plugin-data")
	pluginDataGroup.Use(middleware.JwtAuthMiddleware())
	pluginDataGroup.Use(middleware.TenantMiddleware())
	{
		pluginDataGroup.GET("/:table", controller.ControllerGroupApp.PluginEngineController.GetPluginData)
		pluginDataGroup.POST("/:table", controller.ControllerGroupApp.PluginEngineController.CreatePluginData)
		pluginDataGroup.PUT("/:table/:id", controller.ControllerGroupApp.PluginEngineController.UpdatePluginData)
		pluginDataGroup.DELETE("/:table/:id", controller.ControllerGroupApp.PluginEngineController.DeletePluginData)
	}
}

File: web/src/api/pluginEngine.ts
import request from './request'

export interface InstalledPlugin {
  id: number
  plugin_id: string
  name: string
  version: string
  icon: string
  description: string
  config_json: Record<string, any>
  license_key: string
  status: 'active' | 'disabled'
  installed_at: string
  updated_at: string
}

export interface InstallPluginRequest {
  plugin_id: string
  version: string
  license_key: string
}

const pluginEngineApi = {
  // 获取已安装的插件列表
  getInstalledPlugins() {
    return request.get('/plugin-engine/plugins')
  },

  // 获取单个插件详情
  getPlugin(pluginId: string) {
    return request.get(`/plugin-engine/plugins/${pluginId}`)
  },

  // 插件安装回调（从市场跳回来时调用）
  installCallback(data: InstallPluginRequest) {
    return request.post('/plugin-engine/install-callback', data)
  },

  // 卸载插件
  uninstallPlugin(pluginId: string) {
    return request.delete(`/plugin-engine/plugins/${pluginId}`)
  },

  // 启用插件
  enablePlugin(pluginId: string) {
    return request.put(`/plugin-engine/plugins/${pluginId}/enable`)
  },

  // 禁用插件
  disablePlugin(pluginId: string) {
    return request.put(`/plugin-engine/plugins/${pluginId}/disable`)
  },

  // 升级插件
  upgradePlugin(pluginId: string, newVersion: string, newLicenseKey: string) {
    return request.put(`/plugin-engine/plugins/${pluginId}/upgrade`, {
      new_version: newVersion,
      new_license_key: newLicenseKey
    })
  },

  // 获取插件配置
  getPluginConfig(pluginId: string) {
    return request.get(`/plugin-engine/plugins/${pluginId}/config`)
  },

  // 更新插件配置
  updatePluginConfig(pluginId: string, config: Record<string, any>) {
    return request.put(`/plugin-engine/plugins/${pluginId}/config`, config)
  },

  // 安装示例插件
  installDemo() {
    return request.post('/plugin-engine/demo/install')
  },

  browseMarket(keyword?: string, category?: string) {
    const params: Record<string, string> = {}
    if (keyword) params.keyword = keyword
    if (category) params.category = category
    return request.get('/plugin-engine/market/plugins', { params })
  },

  installFromMarket(pluginId: string, version: string, licenseKey?: string) {
    return request.post('/plugin-engine/install-callback', {
      plugin_id: pluginId,
      version,
      license_key: licenseKey || '',
    })
  },

  checkUpdates() {
    return request.get('/plugin-engine/plugins/check-updates')
  },
}

export default pluginEngineApi
